README.md
# Explicit Framing (Effect of Experiment Framing on Game Data)

This is a repository of materials and data for the explicit framing experiment in the Effect of Experiment Framing on Game Data study. 

## Directory Structure

* `data/` - processed/anonymised data
* `out/` - output of analysis scripts including data and graphs. Logs of script console output is saved here too
* `python/` - python scripts
* `r/` - r scripts
* `design/` - ethics, preregistration documents, etc.
* `final/` - versions of documents that have been officially submitted somewhere
* `materials/` - contains materials that were used in the experiment
* `img/` - screenshots of the game //TODO: add screenshots
* `raw/` - raw data stored here before processing and anonymisation

## Pre-Registration

Pre-registration documents are found in `design/`. The power analysis script used in the pre-registration can be found in `r/`, which generated `out/power-anaysis.Rout`. You can generate this using the following command:

    R CMD BATCH --quiet r/power-analysis.r out/power-analysis.Rout

## Why 16 inputs?

Because our last experiment used the last 16 inputs, and this makes comparison easier, and it will still include a bit under 95% of the data.

From [intrinsic elicitation 2](https://github.com/davidgundry/intrinsic-elicitation-2), we found the difference in duration between conditions (which would primarily be from the game tutorial) to be 92.17 seconds (= 593.62 - 501.45). If we include this time in the time stated to participants of 8 minutes, that gives an average play time of 387 seconds. It means that most players should be playing the main game for a bit over 6 minutes. From the same experiment we found that the average time per input was 15.37 with a standard deviation of 5.07. We found the value of 2 standard deviations above the mean for this group (above which you'd expect only approximately 5% of the data). This was equivalent to 25.5 seconds per input. Over 387 seconds this works out to 15.18 inputs. Rounding gives 15 inputs. However, for consistency with the previous experiment (and because it is so close) we will use 16.

## Data collection

Data collection start at 10:29 on 21/06/2021. 198 participants started the experiment. 27 returned their submissions, 3 timed out, leaving 168 participants who completed the experiment. Data collection ended at 14:26 on 21/06/2021. 161 records retrieved from database meaning there were 7 missing from failed submissions or non-submissions out of those who completed.

Records were retrieved at 30 submissions to check manipulation is plausible (and we're not just getting nothing at all). Processed data is in `first-30.json`. 2 excluded from first 30 because first langauge not english. The first 30 are included in the overall dataset.

168 participants @ £1 per participant on Prolific = £235.20. Average pay per hour = £5.78. Median completion approx 10.33 but with 18 participants taking much more time than reasonable. Participants have been paid.

In anonymisation 5 records were removed because of language reported not being English, resulting in 156 published records inlcuded in `data.json`, `age-gender.csv`, and `duration.csv`.

Excluded due to moves < 16 = 36
Excluded due reporting bugs 13

Total of 107 to be included in hypothesis tests, with 57 in the high framing condition and 50 in the low framing condition.

Raw data has been deleted on 22/06/2021.

## Data Source

Data downloaded from [Restdb.io](https://restdb.io) using the script [get-restdb-data](https://github.com/davidgundry/get-restdb-data).

### Pre-anonymised Data Format

Data will be collected using the non-relational database service [restdb.io](https://restdb.io). All data will be downloaded as a json file.

The expected data format is in `raw/debug-example.json`, it looks like this (IDs are random):

    [ 
        {"_id":"5f118222b0e1d1570001ce11","data":{"gameVersion":"HighFraming","loadTime":1594982440874,"uploadTime":1594982944815,"duration":485.424,"playDuration":480.002,"consent":[true,true,true],"answers":["6","6","6","6","1","2","2","2","2","2","2","nobug","","27","female","english","every-day"],"moves":[["circle","square","triangle"],["filled","empty","green"],["red","green","blue"],["big","empty","circle"],["small","empty","triangle"],["small","empty","triangle"],["filled","empty","green"],["red","green","blue"],["big","triangle","empty"],["small","empty","triangle"],["small","empty","circle"],["red","green","blue"],["empty","filled","green"],["empty","filled","green"],["green","red","blue"]],"moveDurations":[1.5314450000005309,1.915494999999486,1.7831050000004325,5.033134999999675,4.784264999998413,7.050825000000259,1.1063899999990099,2.8494300000002113,9.9504450000004,3.452929999999469,11.851834999999483,79.11336499999923,2.4985599999999977,34.44956000000093,3.0330149999990828]},"version":"3.1.0.explicit-framing-3.0.0.HighFraming","studyID":"daa2e38ef9864764b95f4e545","prolificPID":"6440a9870c404843a195ba4a","sessionID":"501517b49a904139b1508183","uid":176}
    ]

* **_id**: Database record ID
* **data**: Participant data recorded from game (a JSON object)
    * **gameVersion**: `HighFraming` or `LowFraming`
    * **loadTime**: a timestamp at the point the game loads (uses JavaScript `Date.now()`)
    * **uploadTime**: a timestamp at the point the game begins uploading data to the server, which is immediately upon the final questionnaire being submitted (uses JavaScript `Date.now()`)
    * **duration**: seconds between submission of pre-test questionnaire (and start of **tutorial**), and play-end interrupt before post-test questionnaire
* **version**: Game version and condition data was collected from
* **studyID, prolificPID, sessionID**: Values from the Prolific service for particiant payment purposes.
* **uid**: Auto-incrementing datanbase record UID.

For other variables, see the 'Data Format' section below.

### Anonymising Data

To prepare (i.e. anonymise) the raw data run the following command from the project directory. 

    python python/prepare-data_exp3.py

Edit that file to set the source data filename to match the one as downloaded.

The script will write files to disk in the folder "data". It will also write many files to disk in the same directory as the source (raw) data file. Many of these are for sanity checking purposes and should be deleted as they are not fully anonymised. The important files are:

* `data.json` (containing the main data)
* `duration.csv` (associating duration and condition)
* `age-gender.csv` (associating age and gender)

**Note:** The script shuffles the lines of the data file and it doesn't do any JSON parsing, so it is almost certain that the end-of-line commas will be incorrect. This will lead cause errors in later analysis scripts. To fix this, ensure there is a comma at the end of each line of data, except for the last one.

## Data Format

### Data.json

After processing, the data (`data.json`) looks like this:

    [
        {"data":{"gameVersion":"HighFraming","playDuration":480.002,"consent":[true,true,true],"answers":["6","6","6","6","1","2","2","2","2","2","2","nobug","","english","every-day"],"moves":[["circle","square","triangle"],["filled","empty","green"],["red","green","blue"],["big","empty","circle"],["small","empty","triangle"],["small","empty","triangle"],["filled","empty","green"],["red","green","blue"],["big","triangle","empty"],["small","empty","triangle"],["small","empty","circle"],["red","green","blue"],["empty","filled","green"],["empty","filled","green"],["green","red","blue"]],"moveDurations":[1.5314450000005309,1.915494999999486,1.7831050000004325,5.033134999999675,4.784264999998413,7.050825000000259,1.1063899999990099,2.8494300000002113,9.9504450000004,3.452929999999469,11.851834999999483,79.11336499999923,2.4985599999999977,34.44956000000093,3.0330149999990828]},"version":"3.1.0.explicit-framing-3.0.0.HighFraming"}
    ]

* **gameVersion**: Either "HighFraming" or "LowFraming".
* **playDuration**: seconds between start of logged (non-tutorial) levels, and play-end interrupt before post-test questionnaire
* **consent**: Checkbox values to consent questions, in order
* **answers**: Answers to the questions:
    1. Seven (7) Likert scale (`1-5`) answers to the Intrinsic Motivation Inventory: Enjoyment Subscale, in default question order
    2. Four (4) Likert scale (`1-6`) answers to the Play Framing questions (see `materials/questions.md`) in order
    3. What is your first language (`english`/`other`)
    4. How often do you play digital games? (`every-day`/`several-times-a-week`/`about-once-a-week`/`about-once-a-month`/`almost-never`)
    5. Answer to the question "Finally, did you encounter any bugs that may have had an effect on how you played the game? No/Yes" (`nobug`/`bug`)
    6. Description of the bug (optional, string / `null`)
* **moves**: Array of moves attempted by the player (sets of three words inputted, whether or not they trigger an action in the game). These are in order they were selected. Moves are in order attempted.
* **moveDurations**: Array of time taken (in seconds) for each move listed in `moves`.
* **version**: Game version and condition data was collected from

### Age and Gender

`data/age-gender.csv` has records in the format: `age (number entry), gender (female/male/other/prefer-not-to-say)`

    36,female
    19,female
    58,male

This also includes participants who are excluded from the hypothesis tests due to too few inputs. (This also applies to participants who reported bugs). If this were not the case, it would prevent changing the minimum-input threshold after data collection (for example, if it becomes clear that the threshold as set is too high) due to the excluded data already being deleted during the process of anonymisation. Either we might not record all the age/genders (if the threshold is lowered), or we record too many age/genders. As we want to publish `data.json` including particpants excluded due to low input (to allow re-running the analysis with different thresholds), it seems better to record the age/genders that match to this potential superset and accept that the counts may not add up to exactly the same number.

### Duration

`data/duration.csv` has records in the format: `condition (HighFraming/LowFraming), duration in seconds`

    HighFraming,486.424
    LowFraming,485.231
    HighFraming,482.029

Because the duration of play (with tutorial) was controlled at 480 seconds (8 minutes), the variation that is observed here is accounted for by the time spent in menus/questionnaires. To see the time spent while moves were being logged in the main game, see the `playDuration` variable in `data.json`. This `duration` is separated from the rest of the data as (in principle) a player who spent an exceptional amount of time in the questionnaires could have their data identified in combination with Prolfic's log data. (`playDuration` is contained within the play and tutorial time that is fixed to almost exactly 480 seconds, so there is no such threat.)

This also includes participants who are excluded from the hypothesis tests due to too few inputs (or reporting bugs) as above.

## Analysis

In the project directory (for this experiment) run the following commands (on Linux). These create (or overwrite) files in `out/`).

    python python/check_exclusions.py > out/check_exclusions.txt

    python python/create_data_csv.py
    
    python python/hypothesis_test_exp3.py > out/hypothesis_test_exp3.txt

    python python/duration_analysis.py > out/duration_analysis.txt

    python python/age_gender_analysis.py  > out/age_gender_analysis.txt

    python python/popular_orders.py > out/popular_orders.txt

## Output Dataset

The file `out/data.csv` is generated by running the `python/create_data_csv.py` script. This is for sanity checking and accessibility to future research.

* **version** - experiment condition
* **language** - first language (english/other)
* **gaming_frequency** - `every-day`/`several-times-a-week`/`about-once-a-week`/`about-once-a-month`/`almost-never`
* **consent_understand, consent_publication, consent** - three explicit consent checkboxes
* **pf1, pf2, pf3, pf4** - answers to 4 Play Framing questions (see `materials/questions.md`)
* **imi1, imi2, imi3, imi4, imi5, imi6, imi7** - answers to the 7 Intrinsic Motivation Inventory questions
* **bug** - whether or not bug reported (bug/nobug)
* **bugdesc** - optional text description of bug
* **imi_enjoyment** - overall IMI Enjoyment score
* **play_framing** - overall Play Framing score
* **total_moves** - number of moves made in game
* **moves_correct_form** - number of moves made that the game could detect are correctly formed (i.e. no duplicate in category, contains exactly 1 noun). In other words, a move that was potentially successful (assuming there were matching blocks to clear)
* **grammatical_moves_idealised** - number of grammatical moves, if against the idealised grammar described below
* **proportion_of_valid_data_total_idealised** - % grammatical (by idealised grammar) moves overall
* **proportion_of_valid_data_first20_idealised** - % grammatical (by idealised grammar) moves out of first 20
* **proportion_of_valid_data_providing_mechanic_actuations_total_idealised** - % grammatical (by idealised grammar) moves out of moves with correct form
* **moves_correct_form_last16** - number of moves with the correct form (see moves_correct_form) out of last 16 moves
* **grammatical_moves_last16_idealised** - number of grammatical moves out of last 16, if against idealised grammar
* **proportion_of_valid_data_last16_idealised** - % grammatical (by idealised grammar) moves out of last 16 moves
* **proportion_of_valid_data_providing_mechanic_actuations_last16_idealised** - % grammatical (by idealised grammar) out of last 16 moves of correct form
* **time_per_input_from_moveDurations** - average number of seconds taken by a move based on data recorded by game. This excludes time spent outside of the level/in level intro/ending modal dialogues. This is the best measure for considering how efficient the _core mechanic_ is at collecting data because it excludes extraneous interaction time.
* **time_per_input_from_8min** - average number of seconds taken by a move if we consider all the play time (after tutorial) including time the player isn't actively in the level, which was set by the experiment at 8 minutes. This is a fairer measure than moveDurations for the efficiency of the game overall at collecting data, as in practice a game is more than just its core mechanic.

### Idealised Grammar

For the idealised grammar, the following rules apply. The words in the game are grouped into category:

    Size: big, small
    Filledness: empty, filled
    Colour: red, blue, green
    Noun: square, circle, triangle, diamond

A grammatical noun phrase must contain a noun and the following precedence must hold (read `<<` to mean preceeds):

    Size << Filledness << Colour << Noun

And they must satisfy the "correct form" rule below.

Thus "big empty blue triangle" is judged grammatical, but if you rearrange that phrase in any way it is ungrammatical. The phrase "big blue empty triangle" would be judged ungramamtical but it might be an acceptable order for many people. As an idealised grammar, it will not necessarily correspond in all ways to actual usage.

### Correct form / mechanic actuations

The game mechanic can never be succesfully actuated by certain types of input. For example, blocks can never be described by two adjectives of the same type (e.g. "red" and "blue"). The game also requires that exactly 1 noun be selected.


changes.md
Prolific
- How is the study presented, labelled, framed on Prolific itself? Can you check and revise all the language there as well?

High framing condition (I assume this is high experimental framing :)

* Consent Form
    - The Consent form still says "play a puzzle game" at the top: change into “Use an app”
    - Consent form says "the design of "applied games". Task says "You will play a puzzle game." Reword this throughout for high framing: "Our research studies the design of interactive media for collecting user data. This study investigates how the presentation of an interactive app influences the way people engage with it." Task: "You will use an interactive app for 8 minutes. There is a tutorial that will show you how to use it."

>> I see the point about self-selection bias. Do we have to fill in a Prolific consent form and then a separate York consent form? In that case, can we drop the Prolific consent form?

>>>If not, and to avoid the self-selection bias, I would suggest the wording on Prolific for the task and for the consent form is neutral/experimental and talks about an “interactive app”.

 

* Text frames:
    - "Applied Game Experiment" overlay: Change into "Interactive Media Experiment". "Your contribution will help us find ways to collect scientific data with interactive apps. We ask you to engage with this app for 8 minutes ..."
    - "Applied Game Experiment" survey: Relabel this "Interactive Media Experiment", put at the end of data collection, not at the beginning (the "how often do you play digital games?" question evokes a framing.)
    - "Start experiment" -> this is the screen I would try and turn into a big splash separator. Again, the language in high framing still carries lots of gaming language. Rephrase: "You will now see an experimental task designed to identify grammatical preferences. In this experiment, you will interact with the app for 8 minutes while we record the inputs that you make."

* Within the game:
    - Change language: Instead of "good job, you cleared the level." Say "Good job, you cleared the task."
    - Throughout all screens, replace "level" with "task"
    - "Clear 3 adjacent blocks for an extra move" instead of "bonus move"
    - Transition after tutorial: "You've got it! Let's play!" Change into "You've got it! Let's continue"
    - In the level feedback "Level cleared", write "Task completed”, “Task X, "Difficulty multiplier", and "next task". Remove the stars if you can. //TODO: should we replace stars, as otherwise we only change text. Removing stars means we are taking a way a game feature, not just re-describing it.
    - Level 2 and following overlays: It shows "9" and a square shape", I have no idea what that means.
    - "Activity complete": Reword "Thank you for participating in the experimental task."

* Dependent variables, play scale
    - A major issue here is that items 1,2,4 say "game." When I was using the game, etc." Reword items 1,2,4 in both conditions: "When I was using the app", "I would characterise my experience with the app as playing", "Using the app felt like ..."
    - Ask for enjoyment/IMI first, then for play framing, else even asking for play framing sets up demand effects ;).


r/power-analysis.r

library("pwr")
# These calculations are to determine the required sample size for the Effect of Experiment Framing on Game
# Data Experiment, comparing Adjective Game in a high framing and low framing condition.
#
# Two hypotheses will be tested, each independent of the other and with an alpha of 0.05.  The first is that
# players will provide lower accuracy data in the low framing condition. The second is that the players will
# report higher enjoyment in the low framing condition.

# Hypothesis 1 (based on previous experiments)
# A two-tailed Mann-Whitney U test will be used to test whether the distribution of Accuracy
# differs significantly between the high-framing condition than the low framing condition. α = 0.05
alt <- "two.sided"
observedEffectSize <- 1.0005893861300805 # From exploratory analysis of the "Enjoyment and data
                                         # quality in a game for human-subject data collection"
                                         # experiments. We found this decrease in accuracy between
                                         # experiment 1 and experiment 2 in the control condition. One
                                         # main change was whether explicit instuctions for
                                         # grammaticality were given
pwr.t.test(d = observedEffectSize, sig.level = 0.05, power = 0.8, alternative = alt)


# Hypothesis 1 (effect size of interest)
effectSizeOfInterest <- cohen.ES(test = "t", size = "medium")$effect.size
                                         # Because the previous experiment is not directly comparable
                                         # and other factors may have significantly contributed (such as
                                         # history effects), we should still check for medium sized effects.
                                         # This assumes about half of the previously observed effect
                                         # was to do with framing(/instructions)
pwr.t.test(d = effectSizeOfInterest, sig.level = 0.05, power = 0.8, alternative = alt)

# Hypothesis 2
# A two-tailed two-sample t-test will be used to test whether the mean scores of Enjoyment
# are greater in the low-framing condition than the high-framing condition. α = 0.05.

# Previous experiments suggest expected effect sizes might be tiny. While this is an interesting hypothesis
# test it is not the focus of the experiment and not worth the money to do a high-powered test. Therefore
# we will not base the sample size calculation on this hypothesis.


# Hypothesis 3
# This is a manipulation check relating to hypothesis 1. We assume the effect size will be relative to the
# effect measured in that hypothesis so will not base our sample size calculation on this hypothesis.


# Largest sample size is is 63.76561 per group. These sample sizes are based on t-tests, which can be less efficient
# than Wilcoxon signed rank tests. The Asymptotic Relative Efficiency (ARE) of a Wilcoxon test relative to a t-test
# has a lower bound of 0.864, meaning that at worst we require 1/0.864 = 1.1574 times the sample size of a t test (Riffenburgh, 2012).
sampleSize <- 63.76561/0.864
sampleSize

# Which we will round to 74. This makes the sample size in both groups 148.
# However, we expect a number of exclusions. Our previous experiment had a exclusion rate of 0.18.
# We have adjusted the wording on the 'bug' question because we were getting several reports that
# probably had no impact, so we expect that proportion to decrease.
# If we add 20 to our sample to increase to 168 participants, that gives us room for an exclusion
# rate of just under 0.12. Including this in the original sample size means we don't need a complicated stopping rule.

# Final sample size: 168


# References

# R.H. Riffenburgh, 2012. Chapter 18 - Sample Size Estimation and Meta-Analysis, in: Statistics in Medicine (Third Edition),
# Editor(s): R.H. Riffenburgh, Academic Press, pp. 365-391, [Relevant section available at:
# https://www.sciencedirect.com/topics/mathematics/asymptotic-relative-efficiency]


python/get_moves.py
import json
import enum

class Categories(enum.Enum):
    Size = 0
    Filled = 1
    Color = 2
    Noun = 3

def all_moves(filename, correct_form, coded):
    """ Returns moves from file as a single array """
    rawData = _load_data(filename)
    moves = []
    for d in rawData:
        if (d["data"]["gameVersion"] == "HighFraming"):
            for m in d["data"]["moves"]:
                if correct_form:
                    if _correct_form(m):
                        if coded:
                            for i, w in enumerate(m):
                                m[i] = _replace(w)
                        moves.append(m)
                else:
                    if coded:
                        for i, w in enumerate(m):
                            m[i] = _replace(w)
                    moves.append(m)
    return moves

def moves_by_user(filename, condition, correct_form, coded):
    """ Returns moves from file as an array of arrays of moves per user. """
    rawData = _load_data(filename)
    moves = []
    for d in rawData:
        if (d["data"]["gameVersion"] == condition):
            u = []
            for m in d["data"]["moves"]:
                if correct_form:
                    if _correct_form(m):
                        if coded:
                            for i, w in enumerate(m):
                                m[i] = _replace(w)
                        u.append(m)
                else:
                    if coded:
                        for i, w in enumerate(m):
                            m[i] = _replace(w)
                    u.append(m)
            moves.append(u)
        
    return moves

def _load_data(filename):
    rawData = []
    with open(filename) as f:
        content = f.read()
        rawData = json.loads(content)
    return rawData

def _correct_form(array):
    a1, a2, a3, n = 0,0,0,0
    adj1 = ["big", "small"]
    adj2 = ["empty", "filled"]
    adj3 = ["red", "blue", "green"]
    nouns = ["square","circle","triangle","diamond"]
    for word in array:
        if word in adj1:
            a1 += 1
        elif word in adj2:
            a2 += 1
        elif word in adj3:
            a3 += 1
        elif word in nouns:
            n += 1
    return (a1 < 2) and (a2 < 2) and (a3 < 2) and (n == 1)

def _replace(word):
    adj1 = ["big", "small"]
    adj2 = ["empty", "filled"]
    adj3 = ["red", "blue", "green"]
    nouns = ["square","circle","triangle","diamond"]
    if word in adj1:
        return Categories.Size
    if word in adj2:
        return Categories.Filled
    if word in adj3:
        return Categories.Color
    if word in nouns:
        return Categories.Noun


python/check_exclusions.py
## Output the number of users in each condition that are not excluded from analysis

import numpy as np
from load_process_exp3 import load_data, process_data

minimum_moves = 16

dataset = 'data'
print("Analysing dataset", dataset, "\n")
rawData = load_data("data/"+dataset)
df = process_data(rawData)

before_consent = len(df.index)
df = df[df['consent_understand'] == True]
df = df[df['consent_publication'] == True]
df = df[df['consent'] == True]
print("Excluded due to lack of consent (should be 0)", before_consent - len(df.index))

before_move_min = len(df.index)
df = df[df['total_moves'] >= minimum_moves]
print("Excluded due to moves <", minimum_moves, "=", (before_move_min - len(df.index)))

before_nobug = len(df.index)
df = df[df['bug'] == "nobug"]
print("Excluded due reporting bugs", before_nobug - len(df.index))

before_language = len(df.index)
df = df[df['language'] == "english"]
print("Excluded due to language", before_language - len(df.index), "(should be 0 as already excluded)")

highFramingCondition = df[df['version']=='HighFraming']
lowFramingCondition = df[df['version']=='LowFraming']

print("Total count", len(df.index))
print("Count in HighFraming condition", len(highFramingCondition.index))
print("Count in LowFraming condition", len(lowFramingCondition.index))

python/load_process_exp3.py
import pandas as pd
import json

def load_data(filename):
    rawData = []
    with open(filename+".json") as f:
        content = f.read()
        rawData = json.loads(content)
    return rawData

def process_data(rawData):
    data, language, version = [], [], [],
    for d in rawData:
        data.append(d["data"])
        language.append(d["data"]["answers"][11])
        version.append(d["data"]["gameVersion"])
            
    gaming_frequency = []
    c1, c2, c3 = [], [], []
    pf1, pf2, pf3, pf4, pf_value = [], [], [], [], []
    imi1, imi2, imi3, imi4, imi5, imi6, imi7, imi_enjoyment = [], [], [], [], [], [], [], []
    bug, bugdesc = [], []
    grammatical_moves_idealised, moves_correct_form, total_moves = [], [] ,[]
    proportion_of_valid_data_idealised, proportion_of_valid_data_providing_mechanic_actuations_idealised = [], []
    proportion_of_valid_data_first20_idealised = []
    grammatical_moves_last16_idealised, moves_correct_form_last16 = [], []
    proportion_of_valid_data_last16_idealised = []
    proportion_of_valid_data_providing_mechanic_actuations_last16_idealised = []
    time_per_input_from_moveDurations = []
    time_per_input_from_8min = []
    for i, d in enumerate(data):
        gaming_frequency.append(d["answers"][12])
        c1.append(d["consent"][0])
        c2.append(d["consent"][1])
        c3.append(d["consent"][2])

        # Calculate IMI Enjoyment subscale mean and Play Framing mean
        # Scores for questions 3 and 4 of IMI and 4 of Play Framing are reversed
        a = [int(numeric_string) for numeric_string in d["answers"][0:11]]
        pf1.append(a[0])
        pf2.append(a[1])
        pf3.append(a[2])
        pf4.append(a[3])
        imi1.append(a[4])
        imi2.append(a[5])
        imi3.append(a[6])
        imi4.append(a[7])
        imi5.append(a[8])
        imi6.append(a[9])
        imi7.append(a[10])
        imi_value = (a[4] + a[5] + (6-a[6]) + (6-a[7]) + a[8] + a[9] + a[10])/7 # Out of 5, so reverse is 6-x
        imi_enjoyment.append(imi_value)
        pf_mean = (a[0] + a[1] + a[2] + (7-a[3]))/4 # Out of 6, so reverse is 7-x
        pf_value.append(pf_mean)

        bug.append(d["answers"][13])
        bugdesc.append(d["answers"][14])

        # Calculate proportions of valid moves (from total):
        count_gram_idealised = sum([is_grammatical_idealised(a) and correct_form(a) for a in d["moves"]])
        count_gram_first20_idealised = sum([is_grammatical_idealised(a) and correct_form(a) for a in d["moves"][:20]])
        count_correct_form = sum([correct_form(a) for a in d["moves"]])
        count_all = len(d["moves"])
        total_moves.append(count_all)
        grammatical_moves_idealised.append(count_gram_idealised)
        moves_correct_form.append(count_correct_form)
        if (count_all > 0):
            proportion_of_valid_data_idealised.append(count_gram_idealised/count_all)
        else:
            proportion_of_valid_data_idealised.append(0)
        proportion_of_valid_data_first20_idealised.append(count_gram_first20_idealised/20)
        if (count_correct_form > 0):
            proportion_of_valid_data_providing_mechanic_actuations_idealised.append(count_gram_idealised/count_correct_form)
        else:
            proportion_of_valid_data_providing_mechanic_actuations_idealised.append(0)
        # Calculate proportions of valid moves (from last 16):
        count_gram_last_16_idealised = sum([is_grammatical_idealised(a) and correct_form(a) for a in d["moves"][-16:]])
        count_correct_form_last_16 = sum([correct_form(a) for a in d["moves"][-16:]])
        grammatical_moves_last16_idealised.append(count_gram_last_16_idealised)
        moves_correct_form_last16.append(count_correct_form_last_16)
        proportion_of_valid_data_last16_idealised.append(count_gram_last_16_idealised/16)
        if (count_correct_form_last_16 > 0):
            proportion_of_valid_data_providing_mechanic_actuations_last16_idealised.append(count_gram_last_16_idealised/count_correct_form_last_16)
        else:
            proportion_of_valid_data_providing_mechanic_actuations_last16_idealised.append(None)


        #Calculate Time per input
        time_per_input_from_moveDurations.append(sum(d["moveDurations"])/len(d["moveDurations"]))
        if count_all > 0:
            time_per_input_from_8min.append((8*60)/count_all)
        else:
            time_per_input_from_8min.append(0)
    d = { 
            "version": version,
            "language": language,
            "gaming_frequency": gaming_frequency,
            "consent_understand": c1,
            "consent_publication": c2,
            "consent":c3,
            "pf1": pf1, "pf2": pf2, "pf3": pf3, "pf4": pf4,
            "imi1": imi1, "imi2": imi2, "imi3": imi3, "imi4": imi4, "imi5": imi5, "imi6": imi6, "imi7": imi7,
            "bug": bug,
            "bugdesc": bugdesc,
            "imi_enjoyment": imi_enjoyment,
            "play_framing": pf_value,
            "total_moves" : total_moves,
            "moves_correct_form": moves_correct_form,
            "grammatical_moves_idealised": grammatical_moves_idealised,
            "proportion_of_valid_data_total_idealised" : proportion_of_valid_data_idealised,
            "proportion_of_valid_data_first20_idealised" : proportion_of_valid_data_first20_idealised,
            "proportion_of_valid_data_providing_mechanic_actuations_total_idealised": proportion_of_valid_data_providing_mechanic_actuations_idealised,
            "moves_correct_form_last16": moves_correct_form_last16,
            "grammatical_moves_last16_idealised": grammatical_moves_last16_idealised,
            "proportion_of_valid_data_last16_idealised" : proportion_of_valid_data_last16_idealised,
            "proportion_of_valid_data_providing_mechanic_actuations_last16_idealised": proportion_of_valid_data_providing_mechanic_actuations_last16_idealised,
            "time_per_input_from_moveDurations": time_per_input_from_moveDurations,
            "time_per_input_from_8min": time_per_input_from_8min
        }
    df = pd.DataFrame(data=d)
    return df

def is_grammatical_idealised(array):
    a = []
    adj1 = ["big", "small"]
    adj2 = ["empty", "filled"]
    adj3 = ["red", "blue", "green"]
    nouns = ["square","circle","triangle","diamond"]
    for word in array:
        if word in adj1:
            a.append(1)
        if word in adj2:
            a.append(2)
        if word in adj3:
            a.append(3)
        if word in nouns:
            a.append(4)
    return (a[0] < a[1] < a[2]) and has_noun(array)


def correct_form(array):
    a1, a2, a3, n = 0,0,0,0
    adj1 = ["big", "small"]
    adj2 = ["empty", "filled"]
    adj3 = ["red", "blue", "green"]
    nouns = ["square","circle","triangle","diamond"]
    for word in array:
        if word in adj1:
            a1 += 1
        elif word in adj2:
            a2 += 1
        elif word in adj3:
            a3 += 1
        elif word in nouns:
            n += 1
    return (a1 < 2) and (a2 < 2) and (a3 < 2) and (n == 1)

def has_noun(array):
    nouns = ["square","circle","triangle","diamond"]
    for word in array:
        if word in nouns:
            return True
    return False